File: solutions/elixir/prime-factors/1/lib/prime_factors.ex
defmodule PrimeFactors do
  @doc """
  Compute the prime factors for 'number'.

  The prime factors are prime numbers that when multiplied give the desired
  number.

  The prime factors of 'number' will be ordered lowest to highest.
  """
  @spec factors_for(pos_integer) :: [pos_integer]
  def factors_for(number), do: collect_factors(number, [], 2)

  defp collect_factors(1, factors, _), do: Enum.reverse(factors)
  defp collect_factors(n, factors, factor) when rem(n, factor) == 0, do: collect_factors(trunc(n/factor), [factor | factors], factor)
  defp collect_factors(n, factors, factor), do: collect_factors(n, factors, factor+1)
end

File: solutions/erlang/prime-factors/3/src/prime_factors.erl
-module(prime_factors).

-export([factors/1]).

factors(Value) -> factors(Value, 2, []).
factors(1, _, Factors) -> Factors;
factors(Value, Factor, Factors) when Value rem Factor == 0 -> factors(Value div Factor, 2, [Factor | Factors]);
factors(Value, NonFactor, Factors) -> factors(Value, NonFactor + 1, Factors).